### PR TITLE
refactor #363

### PR DESCRIPTION
**Pull request by JoranAngevaare**
Fix a few problems and testing for #363

### CODE REVIEW

File: tests/test_wfsim.py [hunk 1]
```diff
@@ -30,27 +31,31 @@ def test_sim_1T():
...
+        )
 
         st = strax.Context(
             storage=tempdir,
             config=dict(
                 nchunk=2, event_rate=1, chunk_size=1,
                 detector='XENON1T',
                 fax_config=('https://raw.githubusercontent.com/XENONnT/strax_auxiliary_files'
-                            '/36d352580b328ff057b1588b8af8c9a6ed8ae704/sim_files/fax_config_1t.json'),  # noqa
+                            '/a5b92102505d6d0bfcdb563b6117bd4040a93435/sim_files/fax_config_1t.json'),  # noqa
```
Comment (JoranAngevaare) [inline tests/test_wfsim.py:47]: important to get the latest version

File: tests/test_wfsim.py [hunk 2]
```diff
@@ -30,27 +31,31 @@ def test_sim_1T():
...
                 fax_config_override=dict(
                     url_base=("https://raw.githubusercontent.com/XENONnT/strax_auxiliary_files"
-                              "/36d352580b328ff057b1588b8af8c9a6ed8ae704/sim_files/"), ),
+                              "/a5b92102505d6d0bfcdb563b6117bd4040a93435/sim_files/"),
+                    **conf_1t,
```
Comment (JoranAngevaare) [inline tests/test_wfsim.py:51]: and update with the test in the load-resource test

File: tests/test_wfsim.py [hunk 3]
```diff
@@ -101,32 +107,27 @@ def test_sim_nT_basics():
...
-        Clone the repo to dali and type 'pytest' to run. The first run will test simple s1,
-        garfield s2 and noise/afterpulses. The second run will test the s1 spline model"""
-
-    if not straxen.utilix_is_configured():
-        log.warning(f"Utilix is not configured, skipping database-requiring tests!")
-        return
-
+@skipIf(not straxen.utilix_is_configured(), 'utilix is not configured')
+def test_sim_nt_advanced(
+        config = None
```
Comment (JoranAngevaare) [inline tests/test_wfsim.py:112]: We should _always_ test the default config - this wasn't done before

File: tests/test_wfsim.py [hunk 4]
```diff
@@ -101,32 +107,27 @@ def test_sim_nT_basics():
...
+                                                 fax_config='fax_config_nt_sr0_v0.json'
+                                                 )
         st.set_config(dict(nchunk=1, event_rate=1, chunk_size=2,))
 
-        st.set_config({'fax_config_override': dict(s2_luminescence_model='simple',
-                                                   s2_time_model="s2_time_spread around zero",
-                                                   s1_lce_correction_map='XENONnT_s1_xyz_LCE_corrected_qes_MCva43fa9b_wires.json.gz',
-                                                   s1_time_spline='XENONnT_s1_proponly_va43fa9b_wires_20200625.json.gz',
-                                                   s1_model_type='optical_propagation+simple',)})
-
```
Comment (JoranAngevaare) [inline tests/test_wfsim.py:129]: add this in a separate test

File: tests/test_wfsim.py [hunk 5]
```diff
@@ -135,13 +136,52 @@ def test_sim_nT_advanced():
...
+def test_nt_advanced_gas_gap_garfield():
+    config = dict(
+        fax_config_override=dict(
+            s2_luminescence_model='garfield_gas_gap',
+            s2_correction_map="XENONnT_s2_xy_map_v4_210503_mlp_3_in_1_iterated.json",
+            s2_luminescence_gg= "garfield_timing_map_gas_gap_sr0.npy",
+            garfield_gas_gap_map= "garfield_gas_gap_map_sr0.json",
+            se_gain_map="XENONnT_se_xy_map_v1_mlp.json",
+    )
+    )
```
Comment (JoranAngevaare) [inline tests/test_wfsim.py:177]: bunch of older test configurations that we should keep supporting

File: tests/test_wfsim.py [hunk 6]
```diff
@@ -175,8 +212,8 @@ def test_sim_mc_chain():
             fax_config_override=dict(
                 s1_model_type='nest',
                 s2_time_model="s2_time_spread around zero",
-                url_base='https://raw.githubusercontent.com/XENONnT/private_nt_aux_files/master/sim_files',
-                s1_lce_correction_map=["constant dummy", 1, []],
+                # url_base='https://raw.githubusercontent.com/XENONnT/private_nt_aux_files/master/sim_files',
+                # s1_lce_correction_map=["constant dummy", 1, []],
```
Comment (JoranAngevaare) [inline tests/test_wfsim.py:216]:
```suggestion
```
don't know why this was there - this should not have worked anyways

File: wfsim/core/rawdata.py [hunk 7]
```diff
@@ -455,7 +457,7 @@ def digitizer_saturation(data, channel_mask):
 @export
 class RawDataOptical(RawData):
 
-    def __init__(self, config, channels=[], timings=[]):
+    def __init__(self, config, channels=tuple(), timings=tuple()):
```
Comment (JoranAngevaare) [inline wfsim/core/rawdata.py:460]: Having mutable default is an absolutely disastrous idea

File: wfsim/core/s2.py [hunk 8]
```diff
@@ -189,19 +189,15 @@ def get_s2_light_yield(positions, config, resource):
...
-            else:
-                sc_gain = resource.se_gain_map(positions)
+           # calculate it from MC pattern map directly if no "se_gain_map" is given
+           sc_gain = resource.s2_correction_map(positions)
+           sc_gain *= config['s2_secondary_sc_gain']
 
-        # depending on if you use the data driven or mc pattern map for light yield for S2 
-        # the shape of n_photon_hits will change. Mc needs a squeeze
+           # depending on if you use the data driven or mc pattern map for light yield for S2
+           # the shape of n_photon_hits will change. Mc needs a squeeze
```
Comment (JoranAngevaare) [inline wfsim/core/s2.py:200]:
```suggestion
       # depending on if you use the data driven or mc pattern map for light yield for S2
       # the shape of n_photon_hits will change. Mc needs a squeeze
```

File: wfsim/core/s2.py [hunk 9]
```diff
@@ -231,17 +227,13 @@ def get_electron_yield(n_electron, xy_int, z_int, config, resource):
...
-        
-        if ('ext_eff_from_map' not in config):
-            cy = electron_lifetime_correction*config['electron_extraction_yield']
-        else:
-            if config['ext_eff_from_map'] == False:
-                cy = electron_lifetime_correction*config['electron_extraction_yield']
-            else:
-                # Extraction efficiency is g2(x,y)/SE_gain(x,y)
-                cy = config['g2_mean']*resource.s2_correction_map(xy_int)*\
-                    electron_lifetime_correction/resource.se_gain_map(xy_int)
```
Comment (JoranAngevaare) [inline wfsim/core/s2.py:243]: this looked a bit convoluted

File: wfsim/load_resource.py [hunk 10]
```diff
@@ -293,9 +293,20 @@ def __init__(self, config=None):
...
+                elif gf_file_name.endswith('npz'):
+                    # Backwards compatibility from before #363 / #370
+                    s2_luminescence_map = straxen.get_resource(gf_file_name, fmt='npy_pickle')['arr_0']
+                    # Get directly the map for the simulated level
+                    liquid_level_available = np.unique(s2_luminescence_map['ll'])  # available levels (cm)
+                    liquid_level = config['gate_to_anode_distance'] - config['elr_gas_gap_length']  # cm
+                    liquid_level = min(liquid_level_available, key=lambda x: abs(x - liquid_level))
+                    self.s2_luminescence = s2_luminescence_map[s2_luminescence_map['ll'] == liquid_level]
+                else:
+                    raise ValueError(f'{gf_file_name} is of unknown format')
```
Comment (JoranAngevaare) [inline wfsim/load_resource.py:309]: This is why the tests kept failing

File: wfsim/pax_interface.py [hunk 11]
```diff
@@ -93,7 +93,9 @@ class PaxEventSimulator(object):
     Call compute to start the simulation process.
     """
 
-    def __init__(self, config={}):
+    def __init__(self, config=None):
+        if config is None:
+            config = {}
```
Comment (JoranAngevaare) [inline wfsim/pax_interface.py:98]: mutable defaults strike again

**Comment by JYangQi00** [comment]
Hi Joran,

Thank you for checking these tests, I really appreciate it!

**Pull request merged**